### PR TITLE
Fix/old file versions

### PR DESCRIPTION
**Pull request by ezavod**
This adds support and tests for old files with version 1. That is mostly for my personal pleasure since i doubt that there is actual use for files which are over 10 years old.

There are also some cleanups of unused code. Is there any reason why one would only test half of the decimal places from XVG? Using full precision works for all my test files.

~Files with version 1 trigger some warnings `Reading file version 1 with version 5 implementation` maybe it would be good to catch those in tests but i am not that much into `pytest` and do not know how.~
Edit: These warnings are catched.

### CODE REVIEW

**Comment by jbarnoud** [comment]
Thank you for the contribution. It is indeed a not super useful but fun feature to have. Because it is not the most useful feature ever, I may take some time to carry the review. Sorry about that. A comment I can make already, though, is that it would not hurt to document some of the changes. More than what the change do, I'd appreciate to see where they come from (what version of gromacs? what file and what function?). I would not mind some documentation about the test files explaining how they got generated; I should have done it with the other files and I start regretting it.

> There are also some cleanups of unused code. Is there any reason why one would only test half of the decimal places from XVG? Using full precision works for all my test files.

I have no recollection of why I did that. But it make no sense to use only half the precision.

**Comment by ezavod** [comment]
No worries time-wise.
I updated most of the changes with comments, hopefully without making it too verbose. However most of the changes in this PR are just transcribed from GROMACS `enxio.cpp`. Therefor I do not posses a conclusive understanding of inner workings. Sadly GROMACS sources and their commits are sometimes sparsely commented as well. I used GROMACS 2018.4 as reference for implementation, but most of the parts were not changed for several years.
Tests were taken from the first commit to GROMACS regressiontests  repo (see coment in source). Sadly I was not able to compile old versions of GROMACS in the time I had. So I can not say much about way they were generated.

As a side note: Large sections of the version 1 code deal with manipulating `esum` and `eav`. Those values are not covert by tests currently. This is not a huge deal since `edr_to_df` does not expose those values. But might be good to consider some tests in the future.

**Comment by orbeckst** [comment]
I don't know what the plans for this PR are going to be (it's from 2019.... feels like half a lifetime away) but for completeness' sake I also updated it. If nothing else, with new GH actions and code coverage it might be easier to decide if/how to merge it.

**Comment by IAlibay** [comment]
My understanding from recent discussions is that we definitely want to get this PR merged eventually but we probably can't afford it to go in the immediate next release. I'm going to tag this as a core requirement for the release after that (0.7.0).

**Comment by ezavod** [comment]
I'm going to rebase the branch on the latest master. This makes it easier for me to work on this branch.
Additionally, files of version 2, 3, 4 are added (float and double) in order to check even [uncommon](https://github.com/MDAnalysis/panedr/issues/37#issuecomment-1169753321) versions.

**Comment by orbeckst** [comment]
@ezavod thank you for all your work! Do you need a review?

**Comment by ezavod** [comment]
>Do you need a review?

Sure, if there is time at hand, feel free to review this PR.

**Comment by orbeckst** [comment]
No promises ;-) ... I hoped that some of the other devs could also be tempted to look (and @BFedder ).

However, with the reorganization after PR #42 you need to do a manual merge/rebase as far as I can tell. Apologies for that, but I don't think that there was an easy way around it (short of merging your PR before Bjarne's work, which he really needs to make progress on his GSOC project).

**Comment by ezavod** [comment]
> you need to do a manual merge/rebase as far as I can tell. Apologies for that

Thankfully, it worked with a simple automatic rebase since everything just got moved around.

However, after the recent restructuring is was necessary to always reinstall panedr when testing changes. This felt somewhat inconvenient for me. So I added the project directory to the `PATH` in the test script. Now, one can simply change something in panedr and run pytest on the changed version without install.

This should be ready for a review now :-)

**Comment by BFedder** [comment]
This is really cool! Thanks for all your work here, for finding the other file versions, and for fixing the coverage issues we've had along the way!

File: tests/test_edr.py
Comment (BFedder) [inline tests/test_edr.py:82]: This should be `# Testfiles for file version 4`

File: panedr/panedr/__init__.py
Comment (BFedder) [inline panedr/panedr/__init__.py:10]: This could maybe use a comment to say that ENX_VERSION is only needed for the tests and not useful for users

File: panedrlite/panedrlite/__init__.py
Comment (BFedder) [inline panedrlite/panedrlite/__init__.py:10]: This could maybe use a comment to say that ENX_VERSION is only needed for the tests and not useful for users

File: panedrlite/panedrlite/panedr.py
Comment (BFedder) [inline panedrlite/panedrlite/panedr.py:88]: This could maybe use a comment to say that ENX_VERSION is only needed for the tests and not useful for users

File: panedrlite/panedrlite/panedr.py
Comment (BFedder) [inline panedrlite/panedrlite/panedr.py:214]: We are raising ValueErrors for most of the other things that could go wrong, so I'm thinking if it might make sense to do the same here or at least raise a warning when the precision is wrong

**Review by BFedder — COMMENTED** [review]
This is looking very good to me, thanks!

File: tests/test_edr.py
Comment (ezavod) [reply tests/test_edr.py]: Oh, too much copy pasta...

File: panedrlite/panedrlite/panedr.py
Comment (ezavod) [reply panedrlite/panedrlite/panedr.py]: Turns out this is actually dead code.

**Comment by orbeckst** [comment]
@ezavod , would you have time to resolve the conflicts on the PR and integrate it with the new pyedr/panedr? It's been a bit hectic getting the package into the new state and I am sorry that your PR was put on hold for a long time. But now that the dust has cleared we'd really like to get your contribution in.

**Comment by ezavod** [comment]
I've rebased everything on the latest main branch so that there are no merge conflicts anymore. This is now ready for review again.

File: pyedr/pyedr/tests/datafiles.py
Comment (BFedder) [inline pyedr/pyedr/tests/datafiles.py:92]: Could you perhaps add a one-liner to make clear that the integer is the version number?

File: pyedr/pyedr/pyedr.py
Comment (BFedder) [inline pyedr/pyedr/pyedr.py:311]:
```suggestion
        Therefore one has to calculate the difference between frames.
```

**Review by BFedder — APPROVED** [review]
Thank you so much for all the work you've put into this, and thanks for bearing with us as we repetitively uprooted the entire package. This is a wonderful contribution, and aside from two very minor things this LGTM. Unless @jbarnoud, @IAlibay, @orbeckst or @hmacdope have anything else, I'll happily merge it and close the book on this PR which has been so long in the making.

**Review by IAlibay — CHANGES_REQUESTED** [review]
I have a couple of things I'd like to discuss here. Also could I get an estimate of how much all the datafiles increase the size of the package by? We were trying hard to keep it down for the last release so I just want to make sure we're not exceeding the 1 MB tarball siz.e

File: panedr/panedr/__init__.py
Comment (ezavod) [reply panedr/panedr/__init__.py]: There were some problems during the 'panedrlite' phase when I was not able to get things working with an editable pip install. This seems to work now.
I removed the unnecessary commit.

**Comment by ezavod** [comment]
> could I get an estimate of how much all the datafiles increase the size of the package by?

This PR adds 230 KiB of new test data. The overall size of the repository should be around 915 KiB. Therefore, I guess the size of the tarball should be < 1MB.
BTW it felt somewhat clunky to add new test files and deal with the pickled unit dict. Maybe a JSON file would be easier for human consumption and not much larger, if at all. But, since the dicts are pickled now in this PR, I don't feel strongly about this anymore.

File: pyedr/pyedr/pyedr.py
Comment (IAlibay) [inline pyedr/pyedr/pyedr.py:330]: I'm ignoring a lot of the other uncovered pieces of code since it's historical poor test coverage, but here and line 351 seems to be new code. Is there a way to test these branches? Possibly using mock / monkeypatch?

File: panedr/panedr/__init__.py
Comment (IAlibay) [inline panedr/panedr/__init__.py:3]:
```suggestion
import pbr.version
from .panedr import ENX_VERSION, edr_to_df, get_unit_dictionary
```
You might have to provide some more information as to why these are necessary. I can't reproduce E402 errors using flake8 or pylint.

I do get a pylint C0411 (wrong-import-order), but that's easily fixable.

**Review by IAlibay — CHANGES_REQUESTED** [review]
Apologies for the long time between reviews, the last month has been rather hectic.

File: panedr/panedr/__init__.py
Comment (ezavod) [reply panedr/panedr/__init__.py]: That was some left over from times when there were some lines in front to set the path. Removed unnecessary `noqa` statements and reordered imports.

File: pyedr/pyedr/pyedr.py
Comment (ezavod) [reply pyedr/pyedr/pyedr.py:330]: I've added some mock files which add ~5 KB of new test data (+ 5 KB of mock-generation script). Most of the new files are malformed (invalid magic etc) and should trigger most warnings and errors. Also, all of the legacy code paths are covered (and it seems nobody has ever tried to read char or int64 sub blocks).
Some errors/warnings are logically not reachable and were replaced by `assert`s.

**Comment by ezavod** [comment]
I guess this is now ready for another round of reviews :-D

File: pyedr/pyedr/tests/data/mocks/edr_mock.py
Comment (IAlibay) [inline pyedr/pyedr/tests/data/mocks/edr_mock.py:2]: Could you explain this file? It's not immediately clear to me what it does

Comment (ezavod) [reply pyedr/pyedr/tests/data/mocks/edr_mock.py:7]: This is just some helper script to fudge some artificial EDR files with special values. As I'm not aware how to reach some of the code branches with data generated by Gromacs itself, I used this to generate the mocks.
The inclusion of the script is only for archiving/posterity.

Comment (ezavod) [reply pyedr/pyedr/tests/data/mocks/edr_mock.py:7]: Added some comments in the file to document its usage a bit more.

**Comment by IAlibay** [comment]
@ezavod I'm so sorry this has yet again stalled in review, hopefully you're still willing to contribute this PR.

@MDAnalysis/coredevs can someone take over here to give it a secondary review please? Maybe @orbeckst ?

**Comment by ezavod** [comment]
> @ezavod I'm so sorry this has yet again stalled in review, hopefully you're still willing to contribute this PR.

Sure, I'm still happy to move this forward. I've rebased the branch on latest main.

**Comment by IAlibay** [comment]
Pinging @MDAnalysis/coredevs again to over a review here please.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Thanks a lot @ezavod — that's some serious code archaeology and thought into how to test. I can't really comment on the correctness of the details of all the EDR formats — I am going to trust you there.

I saw how you had addressed various comments about adding documentation/comments and this all looks good to me.

The only thing that is missing: Please add an entry to ChangeLog (I think you need to open a 0.8.0 entry). Then I am more than happy to ✅ & ship it .... thank you for your patience!

**Comment by ezavod** [comment]
> Please add an entry to ChangeLog (I think you need to open a 0.8.0 entry)

Done.

Thank you all for the supportive feedback and reviews over the last nearly 5 years ;-)

**Review by orbeckst — APPROVED** [review]
Thanks a lot, nice work.

Not that often that "aged code" is like good wine 🍷 .

**Comment by orbeckst** [comment]
@jbarnoud you're officially assigned as the PR shepherd. Do you want to do the honors?

**Comment by jbarnoud** [comment]
I think I assigned myself 5 years ago when I thought I would find time the next week or so... I'll very gladly merge this. Thank you @ezavod for your PR and for your patience.

**Pull request merged**
